music21/midi/translate.py

```diff
@@ -1787,27 +1795,53 @@ def _prepareStreamForMidi(s):
         s = s.expandRepeats()  # makes a deep copy
     else:
         s = copy.deepcopy(s)
-    if s.hasPartLikeStreams():
-        # check for tempo indications in the score
-        mmTopLevel = s.iter.getElementsByClass('MetronomeMark').stream()
-        if mmTopLevel:  # place in top part
-            target = s.iter.getElementsByClass('Stream')[0]
-            for mm in mmTopLevel:
-                target.insert(mmTopLevel.elementOffset(mm), mm)
-                s.remove(mm)  # remove from Score level
-        # TODO: move any MetronomeMarks not in the top Part to the top Part
 
+    conductor = conductorStream(s)
+
+    if s.hasPartLikeStreams():
         # process Volumes one part at a time
         # this assumes that dynamics in a part/stream apply to all components
         # of that part stream
         # this sets the cachedRealized value for each Volume
         for p in s.iter.getElementsByClass('Stream'):
             volume.realizeVolume(p)
 
+        s.insert(0, conductor)
+        out = s
+
     else:  # just a single Stream
         volume.realizeVolume(s)
+        out = stream.Score()
+        out.insert(0, conductor)
+        out.insert(0, s)
+
+    return out
 
-    return s
+
+def conductorStream(s: stream.Stream) -> stream.Part:
+    '''
+    Strip the given stream of any events that belong in a conductor track
+    rather than in a music track, and return a :class:`~music21.stream.Part`
+    containing just those events, without duplicates.
+    '''
+    from music21 import tempo, meter
+    out = stream.Part()
+
+    condTrkEvents = s.flat.getElementsByClass((
+        'MetronomeMark', 'TimeSignature', 'KeySignature'))
+    for el in condTrkEvents:
+        o = condTrkEvents.elementOffset(el)
+        s.remove(el, recurse=True)
+        # Don't overwrite anything at this offset
+        if not out.getElementsByOffset(o).getElementsByClass(el.classes[0]):
+            out.insert(o, el)
```

Comment:
If this gets slow (like working with a Huge score), this is exactly what tree.trees.OffsetTree() is meant to do extremely quickly. OR since o is sorted and out starts out empty, you can keep track of lastOffset and just do the check if o is the same as lastOffset.  If you keep track of elements at the lastOffset (or just the classes) then you can skip the whole .getElementsByOffset() and .getElementsByClass() and therefore use .coreInsert() with a .coreElementsChanged() afterwards and get a huge speed increase.